src/Arcus.EventGrid/Contracts/EventGridEvent.cs

```diff
@@ -0,0 +1,100 @@
+    /// </summary>
+    /// <typeparam name="TData">Type of data payload</typeparam>
+    public class EventGridEvent<TData> : EventGridEvent, IEvent
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEvent{TData}"/> class.
+        /// </summary>
+        protected EventGridEvent() { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEvent{TData}"/> class.
+        /// </summary>
+        /// <param name="id">The unique identifier of the event.</param>
+        /// <param name="dataVersion">The schema version of the data object. The publisher defines the schema version.</param>
+        /// <param name="eventType">The one of the registered event types for this event source.</param>
+        public EventGridEvent(string id, string dataVersion, string eventType)
+            : base(id, subject: null, data: default(TData), eventType: eventType, eventTime: DateTime.UtcNow, dataVersion: dataVersion)
+        {
+            Validate();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEvent{TData}"/> class.
+        /// </summary>
+        /// <param name="id">The unique identifier of the event.</param>
+        /// <param name="data">The event data specific to the resource provider.</param>
+        /// <param name="dataVersion">The schema version of the data object. The publisher defines the schema version.</param>
+        /// <param name="eventType">The one of the registered event types for this event source.</param>
+        public EventGridEvent(string id, TData data, string dataVersion, string eventType)
+            : base(id, subject: null, data: data, eventType: eventType, eventTime: DateTime.UtcNow, dataVersion: dataVersion)
+        {
+            Validate();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridEvent{TData}"/> class.
+        /// </summary>
+        /// <param name="id">The unique identifier of the event.</param>
+        /// <param name="subject">The publisher-defined path to the event subject.</param>
+        /// <param name="data">The event data specific to the resource provider.</param>
+        /// <param name="dataVersion">The schema version of the data object. The publisher defines the schema version.</param>
+        /// <param name="eventType">The one of the registered event types for this event source.</param>
+        public EventGridEvent(string id, string subject, TData data, string dataVersion, string eventType)
+            : base(id, subject, data: data, eventType: eventType, eventTime: DateTime.UtcNow, dataVersion: dataVersion, metadataVersion: "1")
+        {
+            Validate();
+        }
+
+        /// <summary>
+        ///     Event data specific to the resource provider.
+        /// </summary>
+        public TData GetPayload()
```

Comment:
Let's change this to `Payload` property as discussed

Reply:
But throwing exceptions in a property is not a recommended approach. Therefore I would recommend a method like this.

Reply:
The idea of this class is to make it simple to consume the body and for me a property feels a lot easier as from a consumer perspective they don't care what happens under the hood, we just want the typed payload.

Using a method already feels like another hurdle.

What do you think @fgheysels 	?

Reply:
What should we do?

Reply:
Let's wait for @fgheysels 	his feedback and proceed then, he will give us quorum.

Reply:
I think `GetPayload()` instead of a a `Payload` property is  a valid choice, when some parsing / calculation is required or when exceptions can be thrown.

But actually, I'm wondering what the use of this is:

```csharp
JObject.Parse(base.Data.ToString()).ToObject<TData>()
```
A type is converted to a string, and then the string is parsed again to an object of the same type ?
I assume that you rely here on the fact that the `ToString()` method returns a json representation of the instance ?

The `Data` property is always assigned as a `TData` type , so why are we doing this ?
Or am I missing something here and am I completely wrong ? :)

Reply:
Not exactly parsed to the same type. `base.Data` returns an `object`. So we need to parse the JSON representation of that object to a typed instance.

Reply:
Ok, I see.  
I'd go for a method in this case instead of a property.

Reply:
Method it is!